lib/covalence/core/cli_wrappers/popen_wrapper.rb

```
@@ -77,6 +110,37 @@ def spawn_subprocess(env, run_cmd,
         return wait_thread.value.exitstatus
       end
 
+      def spawn_subprocess_with_workdir(env, run_cmd,
+                       stdin_io: STDIN,
+                       stdout_io: STDOUT,
+                       stderr_io: STDERR,
+                       ignore_exitcode: false,
+                       path: Dir.pwd,
+                       workdir: Dir.pwd)
+        logger.info "path: #{path} workdir: #{workdir} run_cmd: #{run_cmd}"
+        Signal.trap("INT") {} #Disable parent process from exiting, orphaning the child fork below
+        wait_thread = nil
+        prefix=path.gsub(/^\/workspace*/,'')
+        Open3.popen3(env, "prefixout -p '#{prefix} ' -- " + run_cmd, :chdir=>workdir) do |stdin, stdout, stderr, wait_thr|
+          mappings = { stdin_io => stdin, stdout => stdout_io, stderr => stderr_io }
+          wait_thread = wait_thr
+
+          Signal.trap("INT") {
+            Process.kill("INT", wait_thr.pid)
+            Process.wait(wait_thr.pid, Process::WNOHANG)
+            exit(wait_thr.value.exitstatus)
+          } # let SIGINT drop into the child process
+
+          handle_io_streams(mappings, stdin_io)
+        end
+
+        Signal.trap("INT") { exit } #Restore parent SIGINT
+
+        return 0 if ignore_exitcode
+        exit(wait_thread.value.exitstatus) unless wait_thread.value.success?
```

I do not think we should `exit` from any internal method like this, instead I feel we should always return the retval and the parent can decide if they really want to panic like that.

Reasoning:
- Much harder to test added conditional branch complexity of this method
- Doesn't save you much time
- Makes it easier to make hard to debug code
- Allows to simplify this to a single line.
- `ignore_exitcode` would no longer be required, which is great because it's existence means we are *sometimes* eating the error without any possibility of recovering it.

---

commented. It's probably in the next refactoring effort to take a look at the whole Signal.trap and see why we ned to do that for just `ctrl-c`